src/streams.h

```diff
@@ -802,16 +800,19 @@ class CBufferedFile
 
     //! rewind to a given reading position
     bool SetPos(uint64_t nPos) {
-        nReadPos = nPos;
-        if (nReadPos + nRewind < nSrcPos) {
-            nReadPos = nSrcPos - nRewind;
+        size_t bufsize = vchBuf.size();
+        if (nPos + bufsize < nSrcPos) {
+            // rewinding too far, rewind as far as possible
+            nReadPos = nSrcPos - bufsize;
```

Comment:
Would be nice to have a test case for "rewinding too far" in the test `streams_buffered_file` as well, since this code was changed.

Reply:
Done, https://github.com/bitcoin/bitcoin/pull/16577/commits/723baf279ecf42c42c4c73378e6bc175ae07b631.